### ISSUE TITLE
Timezone issues

### ISSUE THREAD

**Issue by petecooper** [textpattern/textpattern#1128]
Placeholder issue. I'm going to combine 4-5 existing issues into one for housekeeping reasons.

**Comment by petecooper** [comment]
Please apply `Component: core` to the labels as per the original issues. Thanks!

**Comment by bloatware** [comment]
It looks like most (all?) dates are stored in db as UTC timestamps, which is great. But we have two timezones interfering: the server default one (if set) and txp `timezone_key` pref. Which results in `strftime()` and `safe_strftime()` (both used in core) producing different results. I wonder why we don't simply set the default timezone to our pref? What is the point of keeping both?

**Comment by Bloke** [comment]
A good question. My understanding - and this is an educated guess - is that our _timezone_key_ is for displaying dates on the public site. So, when we fetch things and wrangle stuff by date in the URL, that's the date which gets displayed: UTC stored date +/- the timezone_key.

The server default timezone is just that: the timezone of the server, which may be physically separate from where your intended audience are. This timezone is used when _storing_ all dates in UTC, and thus applying the correct offset when data is received from a browser that may be in a different timezone to the server.

Example:
* Server in San Francisco (default timezone = GMT-8) which houses three separate website domains:
* example.co.uk = timezone_key = GMT
* example.com = GMT-5 (for argument's sake)
* example.eu = GMT+1

As far as I know, that's why the two `(safe_)strftime` calls return different values, but I may be way off.

**Comment by bloatware** [comment]
Yes, but users date input is expected to respect `timezone_key` zone (they don't even have an idea of what the server tz is). Why wouldn't we just set the default timezone to `timezone_key` and use `strtotime()` to convert dates to UTC directly, rather then calculate the server/pref offset and then juggle with
```
    $ts = strtotime($time_str);

    // tz_offset calculations are expensive
    $tz_offset = tz_offset($ts);

    return strtotime($time_str, time() + $tz_offset) - $tz_offset;
```
I can not see where would we need to input a San Francisco date if our audience is in London.

**Comment by bloatware** [comment]
I have actually tried it, and it worked fine on my localhost, but some articles `year/month/date` URL were not recognized on the demo server.

**Comment by Bloke** [comment]
> Why wouldn't we just set the default timezone to `timezone_key`

Which timezone_key? If hosting multiple domains on a single server, doesn't that mean flipping the server default setting back and forth every request? Won't they clash if multiple users were editing data on the CMS attached to both domains?

Unless I've got my wires crossed. The default timezone you're talking about is the _system_ timezone setting, right? Or do you mean a default timezone that's under our control inside a Txp instance? If the latter then, yes, I agree all the `+tz_offset - tz_offset` nonsense seems superfluous.

**Comment by bloatware** [comment]
Oh, I didn't mean altering `php.ini`, just calling `date_default_timezone_set()`, which seems to act per session, like `setlocale()` etc (it would be too bad otherwise).

Well, there is no real problem to use both server and site timezones, it just needs more attention. A more important issue is db server timezone, which can be different yet. Not sure we take it into account everywhere.

**Comment by Bloke** [comment]
Ah yes, sorry for being dim. In which case, I expect `date_default_timezone_set()` is the winner. That function didn't exist when Txp was born in PHP 4 days, which is probably why we had to use all these workarounds. By modernising and using that function, we could probably save a craptonne of hacks.

**Comment by bloatware** [comment]
Okay, let's try it. I had no issues on localhost, but Pete's server has 2h db/server tz offset, so some articles published near midnight go west. Investigating.

**Comment by petecooper** [comment]
Demo server is UTC, database should follow geolocation (check `phpinfo()`, it's accurate).

**Comment by bloatware** [comment]
Yes, db is UTC, but not php:
```
Server local time: 2021-08-06 00:01:00
...
Time zone (GMT offset in seconds): Europe/Amsterdam (3600)
...
Database server time: 2021-08-05 22:01:00
```
This is not forbidden, per se, but makes that an article published at `2021-08-06 00:01:00` (php time) is saved in db with `2021-08-05 22:01:00` datetime value, which makes its URL invalid in `year/month/date` scheme. Additionally, you must do some rewrites that make it fail even more (I would like to know the details).

The problem will appear on any db server that is in a different tz than php. Individual URLs should be recognized, but search by `2021/08/06/` will fail.

**Comment by petecooper** [comment]
Strange. I haven't changed the database time at all, I just set the server to UTC and then let the other stuff do its own thing with appropriate geo stuff applied. It's all on the same host, so I'm not really sure why there's a two-hour offset.

is this db offset problematic? I'm very happy to change it if you'd like me to, it makes no difference from my point of view. I suppose there's some value in it being a bit…weird.

**Comment by Bloke** [comment]
@petecooper I'd leave it for now. As you say, it's a good test of functionality to check that we can handle situations where the DB is physically separate in a different timezone. There are hosts that do that. Well, splitting the DB to its own server, not so sure about in a different timezone but in theory it's possible.

**Comment by bloatware** [comment]
Db offset is quite legit, but problematic in `year/month/day` URL mode. We must fix it anyway, so please leave it as is.

**Comment by bloatware** [comment]
`SET GLOBAL time_zone = '-02:00'` if you fancy local testing.

**Comment by bloatware** [comment]
A quick fix looks easy via `CONVERT_TZ` or `DATE_ADD` SQL functions, once php/sql timezones are compared, but they do not use the same conventions: what is SQL `SYSTEM` for PHP?
